### PR TITLE
MAINT: Use XOR for bool arrays in `np.diff`

### PR DESCRIPTION
**Pull request by soupault**
The binary `-` operator is no longer supported for booleans. Use xor instead to maintain backwards compatibility.

Added some tests for this case, and also for float.

Closes https://github.com/numpy/numpy/issues/9251.

### CODE REVIEW

File: numpy/lib/function_base.py [hunk 1]
```diff
@@ -1912,10 +1911,16 @@ def diff(a, n=1, axis=-1):
     slice2[axis] = slice(None, -1)
     slice1 = tuple(slice1)
     slice2 = tuple(slice2)
+
+    if a.dtype == bool:
```
Comment (seberg) [inline numpy/lib/function_base.py]: I somewhat prefer `== np.bool_`.

Comment (seberg) [inline numpy/lib/function_base.py]: Also it would be nice if you can make the commit use our typical format such as `BUG: ...` (or maybe `MAINT:` here even, sometimes its hard to say ;)).

Comment (soupault) [inline numpy/lib/function_base.py]: @seberg should be good now.

File: numpy/lib/function_base.py [hunk 2]
```diff
@@ -1912,10 +1911,16 @@ def diff(a, n=1, axis=-1):
     slice2[axis] = slice(None, -1)
     slice1 = tuple(slice1)
     slice2 = tuple(slice2)
+
+    if a.dtype == np.bool_:
+        da = a[slice1] ^ a[slice2]
```
Comment (eric-wieser) [inline numpy/lib/function_base.py]: I might be tempted to use `!=`?

Comment (eric-wieser) [inline numpy/lib/function_base.py]: Furthermore, if we use `!=`, then we can add support for dtypes with fields here too. (Possibly an argument for a different name)

Comment (charris) [inline numpy/lib/function_base.py]: `!=` is kinda nice. Don't know if it add any funtionality. @eric-wieser Could you explain a bit more.

Comment (soupault) [inline numpy/lib/function_base.py]: >we can add support for dtypes with fields here too
@eric-wieser I remember that amazing feature I was always willing to try :). Is better to keep it for a separate PR though.

Comment (eric-wieser) [inline numpy/lib/function_base.py]: > Could you explain a bit more.

Test would become `a.dtype == np.bool_ or a.dtype.fields`, and would enable `np.diff(struct_array)`, meaning `adjacent entries are different.

Of course, it would be weird to not also support this for float arrays, suggesting that the "adjacent entries are different" operation should be given a different name to `diff`, and `diff(bool_array)` should DeprecationWarning, recommending the new name

**Comment by eric-wieser** [comment]
> Could you explain a bit more.

Test would become `a.dtype == np.bool_ or a.dtype.fields`, and would enable `np.diff(struct_array)`, meaning `adjacent entries are different.

Of course, it would be weird to not also support this for float arrays, suggesting that the "adjacent entries are different" operation should be given a different name to `diff`, and `diff(bool_array)` should DeprecationWarning, recommending the new name

**Comment by eric-wieser** [comment]
I think that's outside the scope of this PR though, and should be considered after merging this

File: numpy/lib/function_base.py (hunk 2 above)
Comment (charris) [inline numpy/lib/function_base.py]: I note that at the ufunc level, `logical_xor` is defined as `not_equal` for booleans, but the operator `!=` goes through richcompare while `^` does not, hence `^` is a trifle faster.

One problem with the deprecation of `-` is that if we remove the boolean implementation the bools will be upcast to int8, which will really confuse people, but at least be consistent with python ...

**Comment by charris** [comment]
Thanks @soupault .

**Comment by jakirkham** [comment]
Did this get backported into NumPy 1.13/land in the 1.13.1 release, @charris?

**Comment by charris** [comment]
@jakirkham No, the `-` operator was reinstated with the same old DeprecationWarning.

**Comment by jakirkham** [comment]
Ok, makes sense. Thanks.
